streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -620,22 +620,35 @@ public final TopologyBuilder copartitionSources(Collection<String> sourceNodes)
      * Returns the copartition groups.
      * A copartition group is a group of source topics that are required to be copartitioned.
      *
+     * @param applicationId
      * @return groups of topic names
      */
-    public Collection<Set<String>> copartitionGroups() {
+    public Collection<Set<String>> copartitionGroups(final String applicationId) {
```

Comment:
Good point, I missed this part before. Here are my observations / thoughts:

1. Currently we have three functions in `TopologyBuilder` where we are leaking the application id information into: `sourceTopics`, `sourceTopics` and `copartitionGroups` (added in this patch). All of them are due to the need to add application ids as prefix to these internal topics.

2. There is one idea proposed by @jkreps that we were discussing about, to have "hierarchical topics" in Kafka (https://cwiki.apache.org/confluence/display/KAFKA/Hierarchical+Topics). If this feature is introduced in the future, we will not need to add this prefix anymore for a shared cluster but let each application to work under its own "sub-root" level, and this will likely be done completely in the `StreamPartitionAssignor`, so that `TopologyBuilder` will be totally agnostic to the application id.

3. Within that in mind, personally I'd prefer to keep the application id in the `StreamsConfig`, I'd propose a new solution, which is to add another API function, like `setApplicationId` in `TopologyBuilder` which will be called in `KafkaStreams` constructor by reading from the configs ONLY if `TopologyBuilder` is instance of `KStreamBuilder`, and then the topology builder will remove this parameter from all the above three functions; and moving forward in the future we can get rid of this `setApplicationId` function as well as its logic for appending it as the prefix in the above three functions.

Thoughts?